src/main/java/seedu/task/logic/parser/AddParser.java

```diff
@@ -42,50 +37,42 @@ public AddParser() {}
      */
     @Override
     public Command prepare(String args){
-        final Matcher taskMatcher = TASK_DATA_ARGS_FORMAT.matcher(args.trim());
-        final Matcher eventMatcher = EVENT_DATA_ARGS_FORMAT.matcher(args.trim());
         
-        if (taskMatcher.matches()) {
-            try {
-                Optional <String> dl = getDeadlineFromArgs(taskMatcher.group("deadline"));
-                if (dl.isPresent()) { 
-                    return new AddTaskCommand(
-                        taskMatcher.group("name").trim(),
-                        taskMatcher.group("description").trim(),
-                        dl.get().trim()
-                    );
-                } else {
-                    return new AddTaskCommand(
-                            taskMatcher.group("name").trim(),
-                            taskMatcher.group("description").trim()
-                    );
+        if (args.isEmpty()) {
+            return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
+        }
+        
+        ArgumentTokenizer argsTokenizer = new ArgumentTokenizer(descriptionPrefix, deadlinePrefix, durationPrefix);
+        argsTokenizer.tokenize(args);
+        
+        try {           
+            String name = argsTokenizer.getPreamble().get();
+            Optional <String> description = argsTokenizer.getValue(descriptionPrefix);
+            Optional <String> duration = argsTokenizer.getValue(durationPrefix);
+            Optional <String> deadline = argsTokenizer.getValue(deadlinePrefix);
+            
+            if (duration.isPresent()) { //Only events have duration
+                try {
+                    return new AddEventCommand(name, description.orElse(""), duration.orElse(""));
+                } catch (IllegalValueException ive) {
+                    return new IncorrectCommand(ive.getMessage());
+                } catch (NoSuchElementException nsee) {
+                    return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
```

Comment:
good point. since the handling of the two exceptions are actually identical now.